File: demos/cpp/dsa/trees/heap.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <cassert>

int main() {
    std::cout << "=== C++ std::priority_queue (Binary Heap) ===" << std::endl;
    std::priority_queue<int> max_heap;
    max_heap.push(15);
    max_heap.push(50);
    max_heap.push(30);

    assert(max_heap.top() == 50);
    max_heap.pop();
    assert(max_heap.top() == 30);

    std::priority_queue<int, std::vector<int>, std::greater<int>> min_heap;
    min_heap.push(15);
    min_heap.push(50);
    min_heap.push(10);
    assert(min_heap.top() == 10);

    std::cout << "C++ Priority Queue tests passed successfully." << std::endl;
    return 0;
}

File: demos/python/dsa/trees/heap.py
import heapq

def main():
    print("=== Python heapq Priority Queue ===")
    heap: list[int] = []
    heapq.heappush(heap, 50)
    heapq.heappush(heap, 15)
    heapq.heappush(heap, 30)

    assert heap[0] == 15
    assert heapq.heappop(heap) == 15
    assert heapq.heappop(heap) == 30
    assert heapq.heappop(heap) == 50

    print("Python heapq tests passed successfully.")

if __name__ == "__main__":
    main()
